[WIP] colcontainer: add on-disk implementation of PartitionedQueue

asubiotto opened the pull request:
The Partitioner interface has also been renamed to PartitionedQueue. The
on-disk implementation of a PartitionedQueue is now used by both the external
sorter and joiner, meaning we now have actual disk spilling for these
operators! Note that there are still some items left to take care of stability
wise before disk spilling is production ready.

Release note (sql change): Vectorized sorts (not including sorts with a LIMIT
and an existing ordering on a subset of the ordering columns) and vectorized
hash joins now spill to disk.

Commit:
colcontainer: add on-disk implementation of PartitionedQueue

The Partitioner interface has also been renamed to PartitionedQueue. The
on-disk implementation of a PartitionedQueue is now used by both the external
sorter and joiner, meaning we now have actual disk spilling for these
operators! Note that there are still some items left to take care of stability
wise before disk spilling is production ready.

Release note (sql change): Vectorized sorts (not including sorts with a LIMIT
and an existing ordering on a subset of the ordering columns) and vectorized
hash joins now spill to disk.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45015)
<!-- Reviewable:end -->

asubiotto commented:
The PR cannot be merged as is since when using disk and a memory limit of 1 as most of our tests do, we open an insane amount of partitions (one per batch). We don't yet have code (@yuzefovich is working on this) that will set a maximum number of partitions based on the memory use of each partition and will recursively partition, this PR should probably wait for that.

File descriptor management also sort of depends on recursive partitioning (recursive partitioning gives us a maximum guarantee of open files which can be pre-reserved with a "registry"), but it brings the question into the spotlight of whether tests specifying a memory limit of 1 is useful. Test-wise probably, since it will definitely exercise recursive partitioning and the handling of open file descriptors. However, benchmarks probably want a larger memory limit to test a normal case.

yuzefovich reviewed (approved):
I agree with your points about tests and benchmarks - I just ran into the limitation of using `1` as memory limit when looking at external sorter benchmarks with merging PR.

Overall <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great to see this getting to the finish line!

Another test we should tweak is `TestHashJoinerAgainstProcessor` and `TestSortAgainstProcessor` in `columnar_operators_test.go`.

> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-:-M-vUwPg1POXPzenDVhV:bw61j9t)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

asubiotto commented:
Re memory limit of 1, created a PR: #45144

Commit:
engine/fs: move DeleteDirAndFiles to fs.FS

This is used by the DiskQueue to delete all leftover files in its directory
as well as the directory itself, since DeleteDir fails if the directory has
any children.

Vectorized flows now also implement a hacky RemoveAll on cleanup since there
might be leftover files and directories but DeleteDir fails if so.

Release note: None (code movement)

Commit:
colexec: limit number of open file descriptors using weighted semaphore

The vectorized execution engine could previously open an unlimited number of
files when performing an external sort, join, or when hash routing. A weighted
semaphore is introduced in this commit with a default maximum of 128, which
can be tweaked with the COCKROACH_VEC_MAX_OPEN_FDS environment variable.

To reduce the number of open file descriptors during a sort or a join, it is
now illegal to enqueue to a partition that has been dequeued from. This doesn't
require a change in sorter or joiner behavior and allows the write file
descriptor to be closed as soon as a partition is read from.

Release note: None (feature not present in past release)

Commit:
colexec: rename "auto" to "192auto" and introduce new "auto" setting

vectorize="auto", which would only run streaming operators, now supports
running certain buffering operators that can spill to disk. vectorize="192auto"
retains the semantics of running with vectorize="auto" in 19.2 as an escape
hatch.

Release note (sql change): vectorized sorts and joins are now run by default
when vectorize=auto and the vectorized row count threshold is met. Use
`SET vectorize='192auto'` to revert to the previous behavior.

asubiotto commented:
This PR now also adds support to limiting file descriptors in use by the vectorized engine and changes the `vectorize=auto` setting to mean "run the disk spilling operators as well". `vectorize=192auto` retains the previous behavior. This PR is still a work in progress and depends on the recursive partitioning PRs before it can merge. We should also run some larger-scale tests/modify roachtest before doing so.

Once this PR merges, we'll have disk spilling by default (in the supported operators) in the vectorized engine.

asubiotto commented:
I think that it might be better to separate the commit that enables the disk spilling behavior by default and create a separate PR for that. The current way of limiting file descriptors is to simply acquire the right to do so from a global semaphore. I was initially thinking of pre-reserving a constant amount, but ended up going with requesting these dynamically, which makes it so that we don't need to wait for the recursive partitioning PRs.

yuzefovich reviewed (commented):
> Reviewed 5 of 14 files at r2, 6 of 11 files at r3, 10 of 12 files at r4, 6 of 6 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45015)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/explain_plan.go, line 233 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0QegiiE8NbGhoxW4Ff:-M0QegiiE8NbGhoxW4Fg:bcyz403) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/explain_plan.go#L233)):*
> ```Go
> 		if ctxSessionData.VectorizeMode == sessiondata.VectorizeOff {
> 			isVec = false
> 		} else if !vectorizedThresholdMet && (ctxSessionData.VectorizeMode == sessiondata.Vectorize192Auto || ctxSessionData.VectorizeMode == sessiondata.VectorizeAuto) {
> ```

`VectorizeAuto` should be removed from the condition.
___
*[pkg/sql/colexec/execplan.go, line 126 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0Qf_oJ60tNzNHPqi_5:-M0Qf_oJ60tNzNHPqi_6:bfrfaf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/colexec/execplan.go#L126)):*
> ```Go
> 	MetadataSources        []execinfrapb.MetadataSource
> 	IsStreaming            bool
> 	CanFallBackToDisk      bool
> ```

This field seems to not be set.
___
*[pkg/sql/colexec/routers.go, line 230 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0QgBGzDPXCxVeWzvn6:-M0QgBGzDPXCxVeWzvn7:blnkmby) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/colexec/routers.go#L230)):*
> ```Go
> 	if batch.Length() == 0 {
> 		if o.mu.pendingBatch != nil {
> 			if err := o.mu.data.enqueue(nil, o.mu.pendingBatch); err != nil {
> ```

What is this `nil` argument?
___
*[pkg/sql/colexec/routers_test.go, line 206 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0QgLjFALQgenIEkH4X:-M0QgLjFALQgenIEkH4Y:bssja0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1674c8aa4d955738365087d33ff707525e8a2b94/pkg/sql/colexec/routers_test.go#L206)):*
> ```Go
> 			// also get unblocked with no data other than the zero batch.
> 			unblockEvent: func(_ Operator, o *routerOutputOp) {
> 				o.addBatch(coldata.ZeroBatch, nil /* selection */)
> ```

I find such comments useful.
___
*[pkg/sql/colflow/vectorized_flow.go, line 44 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0Qi6264sEYKE_ax3oC:-M0Qi6264sEYKE_ax3oD:b-h6dmo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/colflow/vectorized_flow.go#L44)):*
> ```Go
> // countingSemaphore is a semaphore that keeps track of the semaphore count from
> // its perspective.
> type countingSemaphore struct {
> ```

Is the only reason for having this counting semaphore is so that we can correctly release the number of FDs used by this flow?
___
*[pkg/sql/colflow/vectorized_flow.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0QiQoSCtsHg3vWi_nM:-M0QiQoSCtsHg3vWi_nN:b-4otmcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/colflow/vectorized_flow.go#L72)):*
> ```Go
> 	operatorConcurrency bool
> 
> 	countingSemaphore *countingSemaphore
> ```

Where is this field instantiated?
___
*[pkg/sql/colflow/vectorized_flow.go, line 185 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0QiYgUCB4bPm7n8_k2:-M0QiYgUCB4bPm7n8_k3:b-5aqjgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/colflow/vectorized_flow.go#L185)):*
> ```Go
> // TODO(asubiotto): Remove once #45098 is resolved.
> func (f *vectorizedFlow) tryRemoveAll(root string) error {
> 	// Unfortunately there is no way to Stat using TempFS, so simply try all
> ```

What is "Stat" here?
___
*[pkg/sql/execinfra/server_config.go, line 137 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0QhSgJ5Rg2qDwtFj_0:-M0QhSgJ5Rg2qDwtFj_1:b-y3cj46) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/execinfra/server_config.go#L137)):*
> ```Go
> 	TempFS fs.FS
> 
> 	// VecFDSemaphore is a weighted semaphore that restricts the number of open
> ```

Why is it "weighted"?
___
*[pkg/sql/sessiondata/session_data.go, line 271 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/45015#-M0Qhj3S3ILYoqgM-5r0:-M0Qhj3TChVU8RryYJnM:b3a53ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/db82a7fce8832ffea579f0e05c8a502b8cc00973/pkg/sql/sessiondata/session_data.go#L271)):*
> ```Go
> 	// operators or buffering operators that can spill to disk will be run using
> 	// columnar execution.
> 	// TODO(asubiotto): Is it an issue that the integer value of this setting is
> ```

I think it might be an issue in the mixed-version cluster. We probably want to bump the cluster version.


<!-- Sent from Reviewable.io -->

asubiotto commented:
TFTR! Sorry for the confusion, this is still a WIP and contains a half-finished commit that I will put in another PR (the cluster setting). Will ping when this is ready for a review.

asubiotto commented:
Closing in favor of another PR (upcoming)
